### PR TITLE
Kernel: Intel Graphics driver improvements before implementing support for future generations

### PR DESCRIPTION
**Pull request by supercomputer7**
These patches are result of dedicated work on supporting the Intel GVT-g emulated device I did in April this year. To accomplish this big goal, many concepts in the Intel driver were needed to be re-implemented, thus this is considered a refactor to some extent.

This doesn't gain support for any newer generation of Intel GPUs besides what is already supported, but it lays good foundations for future work (which hopefully I will be able to do, and if the Intel ARC GPUs will be sold in a fair price, it will surely help me to continue development).

### CODE REVIEW

File: Kernel/Graphics/Intel/Definitions.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * Copyright (c) 2022, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Types.h>
+
+namespace Kernel::IntelGraphics {
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/Definitions.h:12]: Can we squash the commit `Kernel/Graphics: Export Intel definitions to a header file`? 

The definitions in Intel/Definitions.h are moved from NativeDisplayConnector.h in ` Kernel/Graphics: Introduce the IntelDisplayConnectorGroup class` into the DisplayConnectorGroup header, and then moved out into a separate file here in `Kernel/Graphics: Export Intel definitions to a header file `

File: Kernel/Graphics/Intel/DisplayConnectorGroup.h [hunk 2]
```diff
@@ -73,7 +73,6 @@ class IntelDisplayConnectorGroup : public RefCounted<IntelDisplayConnectorGroup>
     void enable_dpll_without_vga(IntelGraphics::PLLSettings const&, size_t dac_multiplier);
     void set_display_timings(DisplayConnector::ModeSetting const&);
     void enable_pipe_a();
-    void set_framebuffer_parameters(size_t, size_t);
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.h]: The commit `Kernel/Graphics: Remove stale method in the Intel driver ` should also be able to be squashed into the commit that creates `Kernel/Graphics/Intel/DisplayConnectorGroup.h`

File: Kernel/Graphics/DisplayConnector.cpp [hunk 3]
```diff
@@ -188,22 +188,24 @@ ErrorOr<void> DisplayConnector::initialize_edid_for_generic_monitor(Optional<Arr
     return {};
 }
 
-void DisplayConnector::set_edid_bytes(Array<u8, 128> const& edid_bytes)
+void DisplayConnector::set_edid_bytes(Array<u8, 128> const& edid_bytes, bool might_be_invalid)
 {
     memcpy((u8*)m_edid_bytes, edid_bytes.data(), sizeof(m_edid_bytes));
```
Comment (ADKaster) [inline Kernel/Graphics/DisplayConnector.cpp:193]: m_edid_bytes shouldn't need this cast right? `it's an unsigned char[128]`. Begs some other questions about why we're using a raw char array and a memcpy here, but that's not particularly relevant to this review :).

File: Kernel/Graphics/Intel/NativeDisplayConnector.cpp [hunk 4]
```diff
@@ -41,25 +41,28 @@ ErrorOr<void> IntelNativeDisplayConnector::create_attached_framebuffer_console(B
...
 }
 
-ErrorOr<void> IntelNativeDisplayConnector::unblank()
+ErrorOr<void> IntelNativeDisplayConnector::set_y_offset(size_t)
 {
     return Error::from_errno(ENOTIMPL);
 }
 
-void IntelNativeDisplayConnector::set_edid_bytes(Badge<IntelDisplayConnectorGroup>, Array<u8, 128> const& edid_bytes)
+ErrorOr<void> IntelNativeDisplayConnector::unblank()
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/NativeDisplayConnector.cpp:64]: `Kernel/Graphics: Prepare Intel driver to support more generations` shuffles set_edid_bytes, set_y_offset and unblank around in this file, which makes the actual diff hard to follow.

File: Kernel/Graphics/Intel/Definitions.h [hunk 5]
```diff
@@ -10,30 +10,14 @@
...
 namespace Kernel::IntelGraphics {
 
-enum class RegisterIndex {
+enum class GlobalGenerationRegister {
     PipeAConf = 0x70008,
     PipeBConf = 0x71008,
-    GMBusData = 0x510C,
-    GMBusStatus = 0x5108,
-    GMBusCommand = 0x5104,
-    GMBusClock = 0x5100,
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/Definitions.h]: I don't understand why in  `Kernel/Graphics: Prepare Intel driver to support more generations` we don't have any enums for GMBus registers. If GMBus IP being at `0x5100` is specific to Gen4, can't we create a `Gen4RegisterIndex` enumeration to hold it?

File: Kernel/Graphics/Definitions.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+#pragma once
+
+#include <AK/Types.h>
+
+namespace Kernel::Graphics {
+
+// Note: Address 0x50 is expected to be the DDC2 (EDID) i2c address.
+static constexpr u8 ddc2_i2c_address = 0x50;
```
Comment (ADKaster) [inline Kernel/Graphics/Definitions.h:14]: Can we squash `Kernel/Graphics: Export DDC I2C address definition to global header file` as well?

Kernel/Graphics/Definitions.h was deleted in ` Kernel/Graphics: Introduce the IntelDisplayConnectorGroup class ` and is resurrected here.

File: Kernel/Graphics/Intel/DisplayConnectorGroup.cpp [hunk 7]
```diff
@@ -249,6 +249,9 @@ ErrorOr<void> IntelDisplayConnectorGroup::set_mode_setting(IntelNativeDisplayCon
 {
     SpinlockLocker locker(connector.m_modeset_lock);
 
+    VERIFY(to_underlying(connector.connector_index()) < m_connectors.size());
+    VERIFY(const_cast<IntelNativeDisplayConnector*>(&connector) == m_connectors[to_underlying(connector.connector_index())].ptr());
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.cpp:246]: I doubt the const cast is required here to compare a pointer to to a const pointer

File: Kernel/Graphics/Intel/NativeDisplayConnector.h [hunk 8]
```diff
@@ -58,8 +72,9 @@ class IntelNativeDisplayConnector final
     // Note: Paravirtualized hardware doesn't require a defined refresh rate for modesetting.
     virtual bool refresh_rate_support() const override { return true; }
 
-    IntelNativeDisplayConnector(IntelDisplayConnectorGroup const&, Type, PhysicalAddress framebuffer_address, size_t framebuffer_resource_size);
+    IntelNativeDisplayConnector(IntelDisplayConnectorGroup const&, ConnectorIndex connector_index, Type, PhysicalAddress framebuffer_address, size_t framebuffer_resource_size);
     Type const m_type { Type::Analog };
+    ConnectorIndex const m_connector_index { 0 };
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/NativeDisplayConnector.h:76]: this should be initialized with `PortA`, rather than 0.

File: Kernel/Graphics/Intel/Transcoder/DisplayTranscoder.cpp [hunk 9]
```diff
@@ -5,17 +5,11 @@
  */
 
 #include <Kernel/Arch/x86/IO.h>
-#include <Kernel/Graphics/Intel/DisplayTranscoder.h>
+#include <Kernel/Graphics/Intel/Transcoder/DisplayTranscoder.h>
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/Transcoder/DisplayTranscoder.cpp:8]: Can we put DisplayTranscoder in the Transcoder/ directory when it's first added? That way we can avoid some feng shui in ` Kernel/Graphics: Prepare to support future Intel generation transcoders `

File: Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+#include <Kernel/PhysicalAddress.h>
+
+namespace Kernel {
+
+enum class GMBusStatus {
+    TransactionCompletion,
+    HardwareReady
+};
+
+enum DisplayPortAuxiliary {
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.cpp]: This enum name feels like it's missing a noun. DisplayPortAuxiliary... what? DisplayPortAuxiliaryCommand? DisplayPortAuxiliaryBusState?

File: Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.cpp [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+IntelGraphics::DisplayPortAuxChannelRegisters volatile& DisplayPortAuxiliaryConnector::aux_channel_by_port_index(PortIndex port_index) const
+{
+    VERIFY(m_access_lock.is_locked());
+    switch (port_index) {
+    case PortIndex::PortA:
+        return *((IntelGraphics::DisplayPortAuxChannelRegisters volatile*)m_all_display_port_aux_registers.base_address().offset(__builtin_offsetof(UnifiedDisplayPortAuxChannelRegisters, set1)).as_ptr());
+    case PortIndex::PortB:
+        return *((IntelGraphics::DisplayPortAuxChannelRegisters volatile*)m_all_display_port_aux_registers.base_address().offset(__builtin_offsetof(UnifiedDisplayPortAuxChannelRegisters, set2)).as_ptr());
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.cpp]: This feels very strange, especially with the c-style cast. Why do we need to cast, and why do we need to use `offsetof` ?

Can we create an object that represents a view onto the "all_aux_registers" mapping for the AuxChannelRegisters that just lets us do `return m_aux_channel_registers->set2` that will automatically do a volatile read?

File: Kernel/Graphics/Intel/DisplayConnectorGroup.cpp [hunk 12]
```diff
@@ -147,12 +147,14 @@ ErrorOr<void> IntelDisplayConnectorGroup::initialize_gen4_connectors()
     // Note: Just assume we will need one Gen4 "transcoder"
     // Note: Main block of registers starting at HorizontalTotalA register (0x60000)
     auto transcoder_registers_paddr = m_mmio_first_region.pci_bar_paddr.offset(0x60000);
+    // Note: Main block of Pipe registers starting at PipeA_DSL register (0x70000)
+    auto pipe_registers_paddr = m_mmio_first_region.pci_bar_paddr.offset(0x70000);
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.cpp]: Is there a reason we can't squash ` Kernel/IntelGraphics: Move pipe management to the Transcoder class ` into the commit that creates CRTDisplayTranscoder?

File: Kernel/Graphics/Intel/DisplayConnectorGroup.h [hunk 13]
```diff
@@ -41,6 +41,11 @@ class IntelDisplayConnectorGroup : public RefCounted<IntelDisplayConnectorGroup>
 private:
     AK_TYPEDEF_DISTINCT_ORDERED_ID(size_t, RegisterOffset);
 
+    enum class AnalogOutputRegisterOffset {
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.h:37]: ` Kernel: Rename GlobalGenerationRegister => AnalogOutputRegisterOffset ` feels like it could be squashed into the initial commit that creates the DisplayConnectorGroup.

File: Kernel/Graphics/Intel/DisplayConnectorGroup.cpp [hunk 14]
```diff
@@ -293,7 +293,7 @@ u32 IntelDisplayConnectorGroup::read_from_general_register(RegisterOffset offset
     return value;
 }
 
-void IntelDisplayConnectorGroup::write_to_analog_output_register(AnalogOutputRegisterOffset index, u32 value) const
+void IntelDisplayConnectorGroup::write_to_analog_output_register(AnalogOutputRegisterOffset index, u32 value)
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.cpp:307]: The IntelDisplayConnectorGroup class is new to this PR, so please squash `Kernel/IntelGraphics: Mark write_to_analog_output_register as non-const` as well.

File: Kernel/Graphics/Intel/DisplayConnectorGroup.cpp [hunk 15]
```diff
@@ -277,7 +277,7 @@ StringView IntelDisplayConnectorGroup::convert_analog_output_register_to_string(
     }
 }
 
-void IntelDisplayConnectorGroup::write_to_general_register(RegisterOffset offset, u32 value) const
+void IntelDisplayConnectorGroup::write_to_general_register(RegisterOffset offset, u32 value)
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.cpp:291]: same comment for `Kernel/IntelGraphics: Mark write_to_general_register as non-const `

File: Kernel/Graphics/Intel/DisplayConnectorGroup.h [hunk 16]
```diff
@@ -25,9 +25,6 @@ class IntelDisplayConnectorGroup : public RefCounted<IntelDisplayConnectorGroup>
     friend class IntelNativeGraphicsAdapter;
 
 public:
-    enum class Generation {
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.h]: another squash comment, can we have Generation in the defintions file at the same time we introduce the concept of Generations? ` Kernel/IntelGraphics: Move Generation enum to the Definitions file `

File: Kernel/Graphics/Intel/DisplayConnectorGroup.cpp [hunk 17]
```diff
@@ -125,20 +125,30 @@ Optional<IntelGraphics::PLLSettings> IntelDisplayConnectorGroup::create_pll_sett
 
 NonnullLockRefPtr<IntelDisplayConnectorGroup> IntelDisplayConnectorGroup::must_create(Badge<IntelNativeGraphicsAdapter>, IntelGraphics::Generation generation, MMIORegion const& first_region, MMIORegion const& second_region)
 {
+    OwnPtr<IntelGraphics::AddressSpace> address_space;
+    {
+        auto paddr = first_region.pci_bar_paddr;
+        // Note: Half the GTTMADDR (PCI BAR 0) is dedicated to MMIO register space, the second half to Virtual memory PTEs, etc.
+        auto address_space_length = first_region.pci_bar_space_length / 2;
+        auto address_space_mapping = MUST(Memory::map_typed<volatile u32>(paddr.offset(address_space_length), address_space_length, Memory::Region::Access::ReadWrite));
+        address_space = MUST(IntelGraphics::AddressSpace::create(generation, move(address_space_mapping)));
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/DisplayConnectorGroup.cpp:134]: if it's really truly required that this address space exist, then we can play games like this with an immediately invoked lambda expression:

```
 NonnullOwnPtr<IntelGraphics::AddressSpace> address_space = [&]()
    {
        auto paddr = first_region.pci_bar_paddr;
        // Note: Half the GTTMADDR (PCI BAR 0) is dedicated to MMIO register space, the second half to Virtual memory PTEs, etc.
        auto address_space_length = first_region.pci_bar_space_length / 2;
        auto address_space_mapping = MUST(Memory::map_typed<volatile u32>(paddr.offset(address_space_length), address_space_length, Memory::Region::Access::ReadWrite));
        return MUST(IntelGraphics::AddressSpace::create(generation, move(address_space_mapping)));
     }();
```

File: Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+#pragma once
+
+#include <AK/Format.h>
+#include <AK/Types.h>
+
+namespace Kernel::IntelGraphics {
+
+class GPUVirtualAddress {
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h:14]: This class looks like it's doing a lot of heavy lifting that it doesn't need to. Could we create it like this instead?
```
class GPUVirtualAddress : public VirtualAddress
{
   using VirtualAddress::VirtualAddress;
};
```
If not... there's a couple more comments.

File: Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        : m_address(address)
+    {
+    }
+
+    [[nodiscard]] constexpr bool is_null() const { return m_address == 0; }
+    [[nodiscard]] constexpr bool is_page_aligned() const { return (m_address & 0xfff) == 0; }
+
+    [[nodiscard]] constexpr GPUVirtualAddress offset(FlatPtr o) const { return GPUVirtualAddress(m_address + o); }
+    [[nodiscard]] constexpr FlatPtr get() const { return m_address; }
+    void set(FlatPtr address) { m_address = address; }
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h:27]: We should not invent a new name for `operator=(FlatPtr)`

File: Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    void mask(FlatPtr m) { m_address &= m; }
+
+    bool operator<=(GPUVirtualAddress const& other) const { return m_address <= other.m_address; }
+    bool operator>=(GPUVirtualAddress const& other) const { return m_address >= other.m_address; }
+    bool operator>(GPUVirtualAddress const& other) const { return m_address > other.m_address; }
+    bool operator<(GPUVirtualAddress const& other) const { return m_address < other.m_address; }
+    bool operator==(GPUVirtualAddress const& other) const { return m_address == other.m_address; }
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/VirtualMemory/GPUVirtualAddress.h:34]: with c++20 all of these can be `= default`

File: Kernel/Graphics/Intel/Plane/G33DisplayPlane.cpp [hunk 21]
```diff
@@ -21,15 +21,16 @@ IntelG33DisplayPlane::IntelG33DisplayPlane(Memory::TypedMapping<volatile PlaneRe
 {
 }
 
-ErrorOr<void> IntelG33DisplayPlane::set_plane_settings(Badge<IntelDisplayConnectorGroup>, IntelGraphics::GPUVirtualAddress aperture_start, PipeSelect pipe_select, size_t horizontal_active_pixels_count)
+ErrorOr<void> IntelG33DisplayPlane::enable(Badge<IntelDisplayConnectorGroup>)
```
Comment (ADKaster) [inline Kernel/Graphics/Intel/Plane/G33DisplayPlane.cpp:24]: Again, can we squash ` Kernel/IntelGraphics: Move DisplayPlane enable code to derived classes ` into the commit that introduces `G33DisplayPlane` ?

File: Kernel/CMakeLists.txt [hunk 22]
```diff
@@ -75,14 +75,14 @@ set(KERNEL_SOURCES
     Graphics/DisplayConnector.cpp
     Graphics/Generic/DisplayConnector.cpp
     Graphics/GraphicsManagement.cpp
+    Graphics/Intel/Auxiliary/DisplayPortAuxiliaryConnector.cpp
+    Graphics/Intel/Auxiliary/GMBusConnector.cpp
```
Comment (ADKaster) [inline Kernel/CMakeLists.txt:79]: All of the files moved in `Kernel/IntelGraphics: Move auxiliary connectors to a separate directory `  are new to this PR, so let's put them there in the initial commit.

**Review by ADKaster — CHANGES_REQUESTED** [review]
These changes overall seem like a good way to split out the files, but there's a lot of commits where early in the PR, we create a new file, and there's two commits later in the file that move things around or move that file, etc. It might be good to scroll through the commits in the per-commit view to verify that the flow of the PR makes sense and doesn't feel like it's creating files just to move half of the declarations later.

**Comment by supercomputer7** [comment]
Let's try again.

File: Kernel/Graphics/Intel/Definitions.h (hunk 5 above)
Comment (supercomputer7) [reply Kernel/Graphics/Intel/Definitions.h]: Because it's not really specific to Gen4. GMBus chips could be in Gen4, Gen6, Gen9 and Gen9.5 of Intel graphics.

**Comment by supercomputer7** [comment]
This is not needed anymore since the work in #17283. Let's close this :)
